ModelBundle/Repository/ContentRepository.php

```diff
@@ -58,9 +58,105 @@ public function findLastPublishedVersionByContentIdAndLanguage($contentId, $lang
      */
     public function findByContentTypeAndChoiceTypeAndKeywordsAndLanguage($language, $contentType = '', $choiceType = self::CHOICE_AND, $keywords = null)
     {
-        $qb = $this->createQueryFindByContentTypeAndChoiceTypeAndKeywordsAndLanguage($language, $contentType, $choiceType, $keywords);
+        $qa = $this->createAggregationQuery();
+        $this->filterPublishedNotDeletedOnLanguage($language);
+
+        $filter1 = $this->generateContentTypeFilter($contentType);
+        $filter2 = $this->generateKeywordsFilter($keywords);
+
+        if ($filter1 && $filter2) {
+            $qa->match($this->appendFilters($filter1, $filter2, $choiceType));
+        } elseif ($filter1) {
+            $qa->match($filter1);
+        } elseif ($filter2) {
+            $qa->match($filter2);
+        } else {
+            return array();
+        }
+
+        $elementName = 'content';
+        $qa->group(array(
+            '_id' => array('contentId' => '$contentId'),
+            'version' => array('$max' => '$version'),
+            $elementName => array('$last' => '$$ROOT')
+        ));
+
+        return $this->hydrateAggregateQuery($qa, $elementName);
+    }
+
+    /**
+     * Filter visible published contents on $language
+     * 
+     * @param string $language
+     */
+    protected function filterPublishedNotDeletedOnLanguage($language)
```

Comment:
How does this work?

You should give the Aggregation query as a parameter of this method!